scenarios/kubernetes_bazel.py

```diff
@@ -50,17 +54,80 @@ def get_version():
     with open('bazel-genfiles/version') as fp:
         return fp.read().strip()
 
+def get_affected(base, pull):
+    """Get affected packages between base sha and pull sha."""
+    diff = check_output(
+        'git', 'diff', '--name-only',
+        '--diff-filter=d', '%s...%s' % (base, pull))
+    return check_output(
+        'bazel', 'query',
+        '--noshow_progress',
+        'set(%s)' % diff).split('\n')
+
+def query(kind, packages):
+    """
+    Run a bazel query against target kind, include targets from args.
+
+    Returns a list of kind objects from bazel query.
+    """
+    return check_output(
+        'bazel', 'query',
+        '--keep_going',
+        '--noshow_progress',
+        'kind(%s, rdeps(//..., set(%s)))' % (kind, ' '.join(packages))
+    ).split('\n')
+
+def clean_file_in_dir(dirname, filename):
+    """Recursively remove all file with filename in dirname."""
+    for parent, _, filenames in os.walk(dirname):
+        for name in filenames:
+            if name == filename:
+                os.remove(os.path.join(parent, name))
+
 def main(args):
     """Trigger a bazel build/test run, and upload results."""
+    # pylint:disable=too-many-branches, too-many-statements, too-many-locals
+    if args.install:
+        for install in args.install:
+            if not os.path.isfile(install):
+                raise ValueError('Invalid install path: %s' % install)
+            check('pip', 'install', '-r', install)
+
     check('bazel', 'clean', '--expunge')
     res = 0
     try:
+        build_pkgs = []
+        test_pkgs = []
+        if args.affected:
+            base = os.getenv('PULL_BASE_SHA', '')
+            pull = os.getenv('PULL_PULL_SHA', '')
+            if not base or not pull:
+                raise ValueError('PULL_BASE_SHA and PULL_PULL_SHA must be set!')
+            affected = get_affected(base, pull)
+            build_pkgs = affected
+            test_pkgs = affected
+
         if args.build:
-            check('bazel', 'build', *args.build.split(' '))
+            build_pkgs += args.build.split(' ')
```

Comment:
ah, I think there was a misunderstanding with my previous comment. I think we want to use `args.build` and `args.test` to filter the `affected` list, rather than adding them to the affected list.

basically, the interface I'm expecting is that I tell the bazel scenario all of the test packages I care about - let's say `//foo/...` and `//bar/...`; I never want to run tests under `//baz/...`.

If I just pass `--test='//foo/... //bar/...'` then I want to run all tests there.

If I do `--test='//foo/... //bar/...' --affected` then it should only run affected tests in `//foo/...` and `//bar/...`, but should not run any tests under `//baz/...`, even if there are changes affecting tests under `//baz/...`.